File: java/ultimate/MessageFormatterFactory.java
package ultimate;

public class MessageFormatterFactory {

    private MessageFormatterFactory() {}
    public static MessageFormatter getMessageFormatterInstance(String messageToHandle) {
        if (messageToHandle.toLowerCase().contains("formatted: json")) {
            return new JsonMessageFormatter();
        }
        if (messageToHandle.toLowerCase().contains("formatted: plain_text")) {
            return new PlainTextMessageFormatter();
        }
        if (messageToHandle.toLowerCase().contains("formatted: xml")) {
            return new XMLMessageFormatter();
        }
        throw new UnsupportedOperationException("failed to format this type of message");
    }
}

File: typescript/patterns/ultimate-design-pattern-challenge/message-formatter-factory.ts
class MessageFormatterFactory {
  private constructor() {}

  static getMessageFormatterInstance(
    messageToFormat: string,
  ): MessageFormatter {
    console.log("messageToFormat: " + messageToFormat);

    switch (messageToFormat.toLocaleLowerCase()) {
      case "json":
        return new JsonMessageFormatter();

      case "xml":
        return new XmlMessageFormatter();

      case "plain_text":
        return new PlainTextMessageFormatter();

      default:
        throw new Error("failed to format this type of message");
    }
  }
}
